src/TorchSharp/Tensor/Tensor.cs

```diff
@@ -881,6 +883,229 @@ public long[] size()
+
+                    using PinnedArray<IntPtr> pinnedArray = new PinnedArray<IntPtr>();
+                    namesRef = (names is null || names.Count() == 0) ? IntPtr.Zero : pinnedArray.CreateArray(dimNamesArray);
+
+                    res = THSTensor_rename_(Handle, namesRef, names is null ? 0 : dimNamesArray.Length);
+                } else {
+
+                    res = THSTensor_rename_(Handle, IntPtr.Zero, 0);
+                }
+
+                if (res == IntPtr.Zero) { torch.CheckForErrors(); }
+                // This is the only situation in which the names change in place.
+                _names = null;
+                return new Tensor(res);
+            }
+
+            /// <summary>
+            /// Refines the dimension names of the input tensor according to names.
+            /// 
+            /// Refining is a special case of renaming that “lifts” unnamed dimensions.A None dim can be refined to have any name; a named dim can only be refined to have the same name.
+            /// Because named tensors can coexist with unnamed tensors, refining names gives a nice way to write named-tensor-aware code that works with both named and unnamed tensors.
+            /// names may contain up to one ellipsis argument, passed as "...". The ellipsis is expanded greedily; it is expanded in-place to fill names to the same length as
+            /// this.shape using names from the corresponding indices of this.names.
+            /// </summary>
+            /// <param name="names">A list of names, one for each dimension in the tensor.</param>
+            /// <remarks>The named tensor API is experimental and subject to change.</remarks>
+            public Tensor refine_names(IEnumerable<string> names)
+            {
+                var dimNamesArray = ExpandEllipsis(names);
+
+                using PinnedArray<IntPtr> pinnedArray = new PinnedArray<IntPtr>();
+                IntPtr namesRef = pinnedArray.CreateArray(dimNamesArray);
+
+                IntPtr res = THSTensor_refine_names(Handle, namesRef, dimNamesArray.Length);
+                if (res == IntPtr.Zero) { torch.CheckForErrors(); }
+                return new Tensor(res);
+            }
+
+            private IntPtr[] ExpandEllipsis(IEnumerable<string?> names)
+            {
+                var namesArray = names.ToArray();
+                IntPtr[] dimNamesArray = new IntPtr[ndim];
+
+                if (namesArray.Contains("...")) {
+                    int idx = -1;
+                    for (int i = 0; i < namesArray.Length; i++)
+                        if (namesArray[i] == "...") { idx = i; break; }
+
+                    if (idx == namesArray.Length - 1) { // '...' appears last.
+
+                        int i = 0;
+                        for (; i < idx; i++)
+                            dimNamesArray[i] = MarshalDimensionString(namesArray[i]);
+
+                        if (has_names()) {
+
+                            var n = this.names.ToArray();
```

Comment:
>var n = this.names.ToArray(); [](http://example.com/codeflow?start=28&length=29)

we can git id of this allocation too I guess.

Reply:
Yes.